overleaf/document-updater
# [misc] migrate to the native mongo driver

### Description

For https://github.com/overleaf/issues/issues/2907
This PR is migrating the mongo driver from mongojs to mongodb -- the native driver.

The current revision is a bit unclean as it mixes Promises with callbacks. Refactoring all of document-updater for async/await is too much work to code and review at this time.

#### Related Issues / PRs

https://github.com/overleaf/issues/issues/2907
Previous migration: https://github.com/overleaf/chat/pull/53 https://github.com/overleaf/chat/pull/54 https://github.com/overleaf/docstore/pull/66 (https://github.com/overleaf/contacts/pull/49 pending in review)

### Review

The acceptance tests are accessing the db directly. I split the PR into the app-part and acceptance tests part to ease the verification of this PR -- e.g. run acceptance tests on the app-commit then on both.

<details>
<summary> changed startup </summary>

The startup behavior has been changed in the chat service already https://github.com/overleaf/chat/pull/54 and is running with success in prod by now.

> This PR changes the startup of the document-updater service to not accept any http requests until after the mongo connection has been established. It also adds the correct way to handle connection setup errors, to catch and log them -- vs just letting them bubble up as unhandled PromiseRejections.
> 
> The web service can start without an established/working mongo connection, which is bad. Eventually we can port this snippet over.

</details>

#### Potential Impact

High. Persisting of changes from redis -- covered by acceptance tests.

#### Manual Testing Performed

Before checking out this branch, make sure the dev-env is in good shape:
- `dev-environment$ bin/up core clsi`
- open a brand new project in the dev-env
- make edits and compile
- flush changes `$ curl 127.0.0.1:3003/flush_all_projects` -- see no failures
- open project again and see the previous changes
- run the app health check
   `$ curl 127.0.0.1:3003/health_check`

In case the flush is failing (which was the case for me), clearing redis and starting over may work.

Checkout this branch and repeat the steps ^.
